qualityattenuation/qualityattenuation.go

```diff
@@ -0,0 +1,133 @@
+		Minimum_latency:           0.0,
+		Maximum_latency:           0.0,
+	}
+}
+
+func (qa *SimpleQualityAttenuation) AddSample(sample float64) {
+	if sample <= 0.0 {
+		// Ignore zero or negative samples because they cannot be valid.
+		// TODO: This should raise a warning and/or trigger error handling.
+		return
+	}
+	qa.Number_of_samples++
+	if sample > qa.Latency_eq_loss_threshold {
+		qa.Number_of_losses++
+		return
+	} else {
+		if qa.Minimum_latency == 0.0 || sample < qa.Minimum_latency {
+			qa.Minimum_latency = sample
+		}
+		if qa.Maximum_latency == 0.0 || sample > qa.Maximum_latency {
+			qa.Maximum_latency = sample
+		}
+		qa.EmpiricalDistribution.Add(sample, 1)
+		qa.Offset_sum += sample - qa.Offset
+		qa.Offset_sum_of_squares += (sample - qa.Offset) * (sample - qa.Offset)
+	}
+}
+
+func (qa *SimpleQualityAttenuation) GetNumberOfLosses() int64 {
+	return qa.Number_of_losses
+}
+
+func (qa *SimpleQualityAttenuation) GetNumberOfSamples() int64 {
+	return qa.Number_of_samples
+}
+
+func (qa *SimpleQualityAttenuation) GetPercentile(percentile float64) float64 {
+	return qa.EmpiricalDistribution.Quantile(percentile / 100)
+}
+
+func (qa *SimpleQualityAttenuation) GetAverage() float64 {
+	return qa.Offset_sum/float64(qa.Number_of_samples) + qa.Offset
+}
+
+func (qa *SimpleQualityAttenuation) GetVariance() float64 {
+	number_of_latency_samples := float64(qa.Number_of_samples) - float64(qa.Number_of_losses)
+	return (qa.Offset_sum_of_squares - (qa.Offset_sum * qa.Offset_sum / number_of_latency_samples)) / (number_of_latency_samples - 1)
+}
+
+func (qa *SimpleQualityAttenuation) GetStandardDeviation() float64 {
+	return math.Sqrt(qa.GetVariance())
+}
+
+func (qa *SimpleQualityAttenuation) GetMinimum() float64 {
+	return qa.Minimum_latency
+}
+
+func (qa *SimpleQualityAttenuation) GetMaximum() float64 {
+	return qa.Maximum_latency
+}
+
+func (qa *SimpleQualityAttenuation) GetMedian() float64 {
+	return qa.GetPercentile(50.0)
+}
+
+func (qa *SimpleQualityAttenuation) GetLossPercentage() float64 {
+	return float64(qa.Number_of_losses) / float64(qa.Number_of_samples)
+}
+
+func (qa *SimpleQualityAttenuation) GetRPM() float64 {
+	return 60 / qa.GetAverage()
+}
+
+func (qa *SimpleQualityAttenuation) GetPDV(percentile float64) float64 {
+	return qa.GetPercentile(percentile) - qa.GetMinimum()
+}
+
+// Merge two quality attenuation values. This operation assumes the two samples have the same offset and latency_eq_loss_threshold, and
+// will return an error if they do not.
+// It also assumes that the two quality attenuation values are measurements of the same thing (path, outcome, etc.).
+func (qa *SimpleQualityAttenuation) Merge(other *SimpleQualityAttenuation) {
+	// Check that offsets are the same
+	if qa.Offset != other.Offset || qa.Latency_eq_loss_threshold != other.Latency_eq_loss_threshold {
+		//"Cannot merge quality attenuation values with different offset or latency_eq_loss_threshold"
```

Comment:
We can have the method return an `error` and you can generate one with this excellent message by calling `fmt.Errorf`.  Again, go convention is that `error` descriptions do not start with uppercase letters -- odd, I know!!